### DISCUSSION TITLE
Getting object properties dynamically

### DISCUSSION THREAD

**Discussion by faveoled** [google/j2cl#178] [Q&A] [answered]
```
@JsType(isNative=true, namespace= JsPackage.GLOBAL)
public class JSON {

    public static native Object parse(String json);
}
```
```
Object parsed = JSON.parse("{\"a\":\"b\"}");
```
Hi. How can I use the `parsed` object to retrieve its properties in Java code?

**Comment by niloc132** [comment, marked as answer]
The https://github.com/google/jsinterop-base/ project provides some tools that can help here - the simplest would be `Js.asPropertyMap(parsed)`, which would return a `jsinterop.base.JsPropertyMap<Object>` instance. This doesn't directly provide a way to read all keys, but you can query known keys as `Object` or as `jsinterop.base.Any` instances.

```java
Object parsed = JSON.parse("{\"a\":\"b\"}");
JsPropertyMap<Object> contents = Js.asPropertyMap(parsed);
// Using Any lets you put in runtime checks for the property type
String value = contents.getAsAny("a").asString();
// You can also directly read the value and cast
String checked = (String) contents.get("a");
String checked2 = Js.cast(contents.get("a"));
String unchecked = Js.uncheckedCast(contents.get("a"));
```

Rule of thumb: Java's cast is for when Java will accept this as correct and you want the type checked, `Js.cast()` is for when Java won't understand (casting String to JsString for example) but it is still correct and can be type checked, and `Js.uncheckedCast()` is for when you want let the runtime accept what might actually fail a type check.

`JsPropertyMap` also provides access to iterate keys:
```java
contents.forEach(key -> {
  // do something with contents.get(key)
});
```


In https://github.com/google/elemental2/ then we also have `elemental2.core.JsObject`, to use various methods on the JS Object type to read properties (consider Object.keys, Object.prototype.hasOwnProperty). Likely not applicable here, but worth knowing about.

**Reply by faveoled** [reply]
Thanks for the detailed answer
